packages/arb-token-bridge-ui/src/hooks/TransferPanel/useGasSummaryStore.ts

```diff
@@ -0,0 +1,234 @@
+  status: GasEstimationStatus
+  estimatedL1GasFees: number
+  estimatedL2GasFees: number
+}
+
+export function useGasSummary(
+  amount: BigNumber,
+  token: TransferPanelSummaryToken | null,
+  shouldRunGasEstimation: boolean
+) {
+  const {
+    app: { arbTokenBridge, isDepositMode }
+  } = useAppState()
+  const networksAndSigners = useNetworksAndSigners()
+  const { l1, l2 } = networksAndSigners
+  const { address: walletAddress } = useAccount()
+  const { gasSummary, setGasSummary } = useGasSummaryStore()
+
+  const l1GasPrice = useGasPrice({ provider: l1.provider })
+  const l2GasPrice = useGasPrice({ provider: l2.provider })
+
+  // Debounce the amount, so we run gas estimation only after the user has stopped typing for a bit
+  const amountDebounced = useDebouncedValue(amount, 1500)
+
+  const [status, setStatus] = useState<GasEstimationStatus>('idle')
+  const [result, setResult] = useState<GasEstimationResult>({
+    // Estimated L1 gas, denominated in Wei, represented as a BigNumber
+    estimatedL1Gas: constants.Zero,
+    // Estimated L2 gas, denominated in Wei, represented as a BigNumber
+    estimatedL2Gas: constants.Zero,
+    // Estimated L2 submission cost is precalculated and includes gas price
+    estimatedL2SubmissionCost: constants.Zero
+  })
+
+  // Estimated L1 gas fees, denominated in Ether, represented as a floating point number
+  const estimatedL1GasFees = useMemo(() => {
+    const gasPrice = isDepositMode ? l1GasPrice : l2GasPrice
+    return parseFloat(utils.formatEther(result.estimatedL1Gas.mul(gasPrice)))
+  }, [result.estimatedL1Gas, isDepositMode, l1GasPrice, l2GasPrice])
+
+  // Estimated L2 gas fees, denominated in Ether, represented as a floating point number
+  const estimatedL2GasFees = useMemo(
+    () =>
+      parseFloat(
+        utils.formatEther(
+          result.estimatedL2Gas
+            .mul(l2GasPrice)
+            .add(result.estimatedL2SubmissionCost)
+        )
+      ),
+    [result.estimatedL2Gas, l2GasPrice, result.estimatedL2SubmissionCost]
+  )
+
+  useEffect(() => {
+    // When the user starts typing, set the status to `loading` for better UX
+    // The value is debounced, so we'll start fetching the gas estimates only when the user stops typing
+    if (shouldRunGasEstimation) {
+      setStatus('loading')
+    }
+  }, [amount, shouldRunGasEstimation])
+
+  useEffect(() => {
+    async function estimateGas() {
+      // Since we are using a debounced value, it's possible for the value to be outdated
+      // Wait for it to sync before running the gas estimation
+      if (!amountDebounced.eq(amount)) {
+        return
+      }
+
+      // Don't run gas estimation if the value is zero or the flag is not set
+      if (amountDebounced.isZero() || !shouldRunGasEstimation) {
+        setStatus('idle')
+        return
+      }
+
+      if (!walletAddress) {
+        return
+      }
+
+      try {
+        setStatus('loading')
+
+        if (isDepositMode) {
```

Comment:
This is a bit hard to follow (transfer panel vibes 😄 ). Should we do something to reuse the method, something like:

```typescript
let estimateGasFunction

if (isDepositMode) {
  estimateGasFunction = token
    ? depositTokenEstimateGas
    : depositEthEstimateGas
} else {
  estimateGasFunction = token
    ? withdrawTokenEstimateGas
    : withdrawEthEstimateGas
}

const estimateGasResult = await estimateGasFunction({
  amount: amountDebounced,
  address: walletAddress,
  l1Provider: l1.provider,
  l2Provider: l2.provider,
  erc20L1Address: token?.address ? token.address : ''
})
```

Reply:
yea, i was debating whether to refactor it because this is copy-pasted
will see what i can do here or move it to another PR